haskell/doctest.bzl

```diff
@@ -0,0 +1,213 @@
+        args.add([
+          "-optl-l{0}".format(lib_name),
+          "-optl-L{0}".format(paths.dirname(lib.path)),
+        ])
+      else:
+        args.add([
+          "-l{0}".format(lib_name),
+          "-L{0}".format(paths.dirname(lib.path)),
+        ])
+
+  header_files = lib_info.header_files if lib_info != None else bin_info.header_files
+
+  sources = set.to_list(
+    lib_info.source_files if lib_info != None else bin_info.source_files
+  )
+
+  all_modules = set.to_list(
+    lib_info.exposed_modules if lib_info != None else bin_info.modules
+  )
+
+  selected_modules = []
+
+  if ctx.attr.modules:
+    for m in all_modules:
+      if m in ctx.attr.modules:
+        selected_modules.append(m)
+  else:
+    selected_modules = all_modules
+
+  args.add(selected_modules)
+
+  ctx.actions.run_shell(
+    inputs = depset(transitive = [
+      depset(sources),
+      set.to_depset(build_info.package_confs),
+      set.to_depset(build_info.package_caches),
+      set.to_depset(build_info.interface_files),
+      set.to_depset(build_info.dynamic_libraries),
+      set.to_depset(header_files),
+      set.to_depset(external_libs),
+      depset(
+        toolchain.doctest
+        + [hs.tools.cat, hs.tools.ghc]),
+    ]),
+    outputs = [doctest_log],
+    mnemonic = "HaskellDoctest",
+    progress_message = "HaskellDoctest {}".format(ctx.label),
+    command = """
+    {doctest} "$@" > {output} 2>&1 || rc=$? && cat {output} && exit $rc
+    """.format(doctest = toolchain.doctest[0].path, output = doctest_log.path),
+    arguments = [args],
+    # NOTE It looks like we must specify the paths here as well as via -L
+    # flags because there are at least two different "consumers" of the info
+    # (ghc and linker?) and they seem to prefer to get it in different ways
+    # in this case.
+    env = dicts.add({
+      "LD_LIBRARY_PATH": get_external_libs_path(external_libs),
+      },
+      hs.env,
+    ),
+  )
+  return doctest_log
+
+def _haskell_doctest_impl(ctx):
+
+  logs = []
+
+  for dep in ctx.attr.deps:
+    logs.append(_haskell_doctest_single(dep, ctx))
+
+  return DefaultInfo(
+    files = depset(logs)
+  )
+
+haskell_doctest = rule(
+  _haskell_doctest_impl,
+  attrs = {
+    "deps": attr.label_list(
+      doc = "List of Haskell targets to lint.",
+    ),
+    "doctest_flags": attr.string_list(
+      doc = "Extra flags to pass to doctest executable.",
+    ),
+    "modules": attr.string_list(
+      doc = """List of names of modules that will be tested. If the list is
+omitted, all modules provided by `deps` will be tested.
+"""),
+  },
+  toolchains = [
+      "@io_tweag_rules_haskell//haskell:toolchain",
+      "@io_tweag_rules_haskell//haskell:doctest-toolchain",
+  ],
+)
+"""Run doctest test on targets in `deps`.
+
+Note that your toolchain must be equipped with `doctest` executable, i.e.
+you should specify location of the executable using the `doctest` attribute
+when you invoke `haskell_toolchain`.
```

Comment:
It's now configured by `doctest_toolchain` rather than `haskell_toolchain`.  (Sorry, looks like I forgot to update this doc when I made that change earlier.)